src/envoy/http/error_translator/filter.cc

```diff
@@ -0,0 +1,211 @@
+    ENVOY_LOG(debug, "Translating HTTP espv2 error headers");
+
+    // Store a copy and scrub details if needed.
+    error_ = utils::getErrorFilterState(
+        *encoder_callbacks_->streamInfo().filterState());
+    if (config_->shouldScrubDebugDetails()) {
+      error_.clear_details();
+    }
+
+    // Translate to JSON.
+    error_json_ = errorToJson(error_);
+
+    // Update content headers.
+    headers.setContentLength(error_json_.size());
+    headers.setReferenceContentType(
+        Envoy::Http::Headers::get().ContentTypeValues.Json);
+
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (is_grpc_response_) {
+    // We keep the response body as is, but modify the filter state.
+    // This will be used by the access log (SC Report).
+    // This handles the trailers-only response by `sendLocalReply`.
+    ENVOY_LOG(debug,
+              "Translating gRPC non-espv2 (upstream envoy filter) "
+              "trailers-only error response");
+
+    if (headers.GrpcStatus() != nullptr) {
+      const absl::string_view grpc_status =
+          headers.GrpcStatus()->value().getStringView();
+      uint32_t parsed;
+      if (absl::SimpleAtoi(grpc_status, &parsed)) {
+        error_.set_code(parsed);
+      } else {
+        error_.set_code(Envoy::Grpc::Status::WellKnownGrpcStatus::Internal);
+      }
+    } else {
+      error_.set_code(Envoy::Grpc::Status::WellKnownGrpcStatus::Internal);
+    }
+    if (headers.GrpcMessage() != nullptr) {
+      const std::string grpc_message =
+          std::string(headers.GrpcMessage()->value().getStringView());
+      error_.set_message(grpc_message);
+    }
+
+    // Store in filter state.
+    utils::setErrorFilterState(*encoder_callbacks_->streamInfo().filterState(),
+                               error_);
+
+    return FilterHeadersStatus::Continue;
+  }
+
+  // Now we are handling `sendLocalReply` from non-espv2 filters.
+  // We need to modify the response body and filter state.
+  // TODO(nareddyt): Do we also need to check the content type is "plain/text"?
+  ENVOY_LOG(debug,
+            "Translating HTTP non-espv2 (upstream envoy filter) error headers");
+
+  // Stop iteration because we will modify some headers based on the body
+  // (example: content length). Expect encodeData to unblock this.
+  headers_ = &headers;
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterDataStatus Filter::encodeData(Envoy::Buffer::Instance& body,
+                                    bool end_stream) {
+  if (isUpstreamResponse()) {
+    // Do not translate any response sent from the upstream.
+    return FilterDataStatus::Continue;
+  }
+
+  if (is_grpc_response_) {
+    // Do not translate body for gRPC. It is not used.
+    return FilterDataStatus::Continue;
+  }
+
+  if (isEspv2FilterError()) {
+    ENVOY_LOG(debug, "Translating HTTP espv2 error body");
+    GOOGLE_DCHECK(!error_.message().empty());
+    GOOGLE_DCHECK(!error_json_.empty());
+
+    if (!end_stream) {
```

Comment:
end_stream should be always true.  We can assert it.

Reply:
Done